crates/arrow_tools/src/record_batch.rs

```diff
@@ -0,0 +1,267 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use arrow::{
+    array::{
+        new_null_array, Array, Int32Array, Int64Array, LargeStringArray, RecordBatch, StringArray,
+        TimestampMicrosecondArray, TimestampMillisecondArray, TimestampSecondArray,
+    },
+    datatypes::{DataType, Field, SchemaRef, TimeUnit},
+};
+use snafu::prelude::*;
+use std::{num::TryFromIntError, sync::Arc};
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Error converting record batch: {}", source))]
+    ConvertRecordBatch { source: arrow::error::ArrowError },
+
+    #[snafu(display("Unable to convert array: {}", source))]
+    ParseTimestampFromString { source: chrono::ParseError },
+
+    #[snafu(display("Unable to convert {} array from {:?} to {:?}", name, from, to))]
+    ConvertArrowArray {
+        from: DataType,
+        to: DataType,
+        name: String,
+    },
+
+    #[snafu(display("Unable to try from int: {}", source))]
+    ConvertInt { source: TryFromIntError },
+
+    #[snafu(display("Unable to use null for field {field}"))]
+    UnableToUseNullForField { field: String },
+
+    #[snafu(display("Unable to downcast array {field} to {downcast_to}"))]
+    UnableToDowncast { field: String, downcast_to: String },
+}
+
+/// Cast a given record batch into a new record batch with the given schema.
+///
+/// # Errors
+///
+/// This function will return an error if the record batch cannot be casted.
+#[allow(clippy::needless_pass_by_value)]
+pub fn try_cast_to(record_batch: RecordBatch, schema: SchemaRef) -> Result<RecordBatch> {
+    let existing_schema = record_batch.schema();
+
+    if schema.contains(&existing_schema) {
```

Comment:
What are the semantics for `.contains()` and why we're returning a new record batch with the passed in schema instead of just returning it directly? It would be good to add this as a comment given its not obvious.

Reply:
Is about each field is a superset of the original one. Mostly it’s about nullable, when casting from a non-nullable to a nullable one, the change on schema needs to be considered.

Reply:
Will add a comment in.